test/lit.cfg

```diff
@@ -856,37 +856,60 @@ elif run_os in ['linux-gnu', 'linux-gnueabihf', 'freebsd', 'windows-cygnus', 'wi
+        return kwargs[run_cpu]
+
+    ndk_platform_tuple = get_architecture_value(armv7="armeabi-v7a",
+                                                aarch64="arm64-v8a")
+    ndk_platform_triple = get_architecture_value(armv7="arm-linux-androideabi",
+                                                 aarch64="aarch64-linux-android")
+    toolchain_directory_name = "{}-{}".format(ndk_platform_triple, config.android_ndk_gcc_version)
+    tools_directory = make_path(config.android_ndk_path, "toolchains",
+        toolchain_directory_name, "prebuilt", "linux-x86_64",
+        ndk_platform_triple, "bin")
     lit_config.note("Testing Android " + config.variant_triple)
     config.target_object_format = "elf"
     config.target_dylib_extension = "so"
+    if run_cpu == "armv7" or run_cpu == "armv7s" or run_cpu == "armv7k":
+        config.target_swiftmodule_name = "arm.swiftmodule"
+        config.target_swiftdoc_name = "arm.swiftdoc"
+    elif run_cpu == "arm64" or run_cpu == "aarch64":
+        config.target_swiftmodule_name = "arm64.swiftmodule"
+        config.target_swiftdoc_name = "arm64.swiftdoc"
     config.target_runtime = "native"
     config.target_swift_autolink_extract = inferSwiftBinary("swift-autolink-extract")
     config.target_sdk_name = "android"
     android_linker_opt = "-L {libcxx} -L {libgcc}".format(
         libcxx=make_path(config.android_ndk_path,
-                         "sources", "cxx-stl", "llvm-libc++", "libs",
-                         "armeabi-v7a"),
+                         "sources", "cxx-stl", "llvm-libc++", "libs", ndk_platform_tuple),
         libgcc=make_path(config.android_ndk_path,
-                         "toolchains",
-                         "arm-linux-androideabi-{}".format(config.android_ndk_gcc_version),
-                         "prebuilt", "linux-x86_64", "lib", "gcc",
-                         "arm-linux-androideabi",
+                         "toolchains", toolchain_directory_name, "prebuilt",
+                         "linux-x86_64", "lib", "gcc", ndk_platform_triple,
                          "{}.x".format(config.android_ndk_gcc_version)))
+    # Since NDK r14 the headers are unified under $NDK_PATH/sysroot, so the -sdk
+    # switch is not enough. Additionally we have to include both the unified
+    # sysroot, and the architecture sysroot.
+    android_include_opt = "-I {sysroot} -I {sysroot_arch}".format(
+        sysroot=make_path(config.android_ndk_path, "sysroot", "usr", "include"),
+        sysroot_arch=make_path(config.android_ndk_path, "sysroot", "usr",
+                              "include", ndk_platform_triple))
     config.target_build_swift = (
-        '%s -target %s -sdk %r %s -Xlinker -pie %s %s %s %s %s'
-        % (config.swiftc, config.variant_triple, config.variant_sdk,
-           android_linker_opt, resource_dir_opt, mcp_opt,
-           config.swift_test_options,
+        '%s -target %s -sdk %r -tools-directory %r %s %s -use-ld=%s %s %s %s %s %s'
```

Comment:
You are dropping the `-pie` flag.  Is that intentional?

Reply:
Yes, `target_build_swift` is used both for building executables and libraries. In the case of building libraries, `-pie` cannot be there because it only applies to executables (and conflicts with `-shared`).

From what I understand, since all Swift code should be PIC/PIE (I think they are basically the same, correct me if I’m wrong), those parameters should be passed internally (which I think it is not happening, because `IRGen/pic.swift` is still failing for Android targets).